### ISSUE TITLE
LSTM many to many resetting policy problem

### ISSUE THREAD

**Issue by mininaNik** [fchollet/keras#2825]
I want to use RNN(LSTM) to predict multiple future values by using multiple historical (previous) values . I have a time series which is shown in the following plot:
![temp](https://cloud.githubusercontent.com/assets/17245428/15577627/97317b3a-235d-11e6-9ad9-609407548453.png)
Indeed I wan to predict '_nxt_time_' in future in '_nxt_step_number_' step numbers with equal distance by using '_prev_time_' previous values in '_prev_step_number_' step numbers. 

My implementation is as follows:

```
 nxt_time = 1000 
 nxt_step_number = 5
 prev_time = 10000
 prev_step_number = 500
 batch_size = 500
 epochs = 30
 hidden_neurons = 50

 (X_train, y_train), (X_test, y_test) = train_test_split(data)
 model = Sequential()
 model.add(LSTM(hidden_neurons,
           batch_input_shape=(batch_size, prev_step_number, 1),
           forget_bias_init='one',
           return_sequences=False,
           stateful=True))
  model.add(Dense(hidden_neurons))
  model.add(Dropout(0.2))
  model.add(RepeatVector(nxt_step_number))
  model.add(LSTM(hidden_neurons,
           batch_input_shape=(batch_size, prev_step_number, 1),
           forget_bias_init='one',
           return_sequences=True,
           stateful=True))
  model.add(TimeDistributed(Dense(1)))
  model.compile(loss='mse', optimizer='rmsprop')

 print('Training')
 numIteration = len(X_train)/batch_size
 for i in range(epochs):
     print('Epoch', i, '/', epochs)
     for j in range(numIteration):
        print('Batch', j, '/',numIteration,'Epoch', i)
        model.train_on_batch(X_train[j*batch_size:j*batch_size+batch_size,], y_train[j*batch_size:j*batch_size+batch_size,])
    model.reset_states()


 print('Predicting')
 predicted_output = model.predict(X_test, batch_size=batch_size)

########

 def train_test_split(df, test_size=0.3):  
    ntrn = int(round(len(df) * (1 - test_size)))
    X_train, y_train = _load_data(df[0:ntrn])
    X_test, y_test = _load_data(df[ntrn:])
    return (X_train, y_train), (X_test, y_test)

  def _load_data(data):  
      prev_interval_length = prev_time/prev_step_number
      nxt_interval_length = nxt_time/nxt_step_number
      cols = len(data[0])
      rows = len(data)-(nxt_time+prev_time)
      docX = np.empty([rows,prev_step_number,cols]) 
      docY = np.empty([rows,nxt_step_number,cols])

      for i in range(len(data)-(nxt_time+prev_time)):
          for j in range(prev_step_number):
              docX[i,j,] = data[i+prev_interval_length*(j),].reshape([1,cols])
         for j in range(nxt_step_number):
               docY[i,j,] = data[i+(prev_interval_length*prev_step_number)+nxt_interval_length*(j),].reshape([1,cols])

     return docX, docY
```

If I reset the states after each training epoch, then sth close to a constant value  is predicted for all 5 future values, But if I don't reset it then the predicted values increase but don't follow the right patterns. In the following figures you can find the plots for 5 future predicted values separately vs their real values. 
![1_0 1](https://cloud.githubusercontent.com/assets/17245428/15580502/4767dfb6-2369-11e6-8610-bd3c816d7046.png)

![1_1 1](https://cloud.githubusercontent.com/assets/17245428/15580432/05c280de-2369-11e6-8787-7ead25a62051.png)
![1_2 1](https://cloud.githubusercontent.com/assets/17245428/15580520/544361ba-2369-11e6-9293-8d8c42e041a6.png)

![1_3 1](https://cloud.githubusercontent.com/assets/17245428/15580440/0da44a26-2369-11e6-80f7-d5525f8c1890.png)
![1_4 1](https://cloud.githubusercontent.com/assets/17245428/15580445/10a98178-2369-11e6-94d6-9f9cea2947e6.png)

Then I started to use stateless mode. I changed my model as follows:

```
model = Sequential()
model.add(LSTM(hidden_neurons,
           input_shape=(prev_step_number, 1),
           forget_bias_init='one',
           return_sequences=False))
  model.add(Dense(hidden_neurons))
  model.add(Dropout(0.2))
  model.add(RepeatVector(nxt_step_number))
  model.add(LSTM(hidden_neurons,
           forget_bias_init='one',
           return_sequences=True))
  model.add(TimeDistributed(Dense(1)))
  model.compile(loss='mse', optimizer='rmsprop')
  model.fit(X_train,y_train,batch_size = batch_size, nb_epoch = epochs, validation_data = (X_valid,y_valid),shuffle = False)
```

 In the following figures you can find the plots for 5 future predicted values separately vs their real values by using this model. These plots show that always we can predict the first future value with high accuracy in stateless mode, although the second ones' pattern is similar to the first ones but I don't know why  the prediction accuracy starts to decrease (the predicted values start to shrink)!! Could anyone help me to fix this problem? What is the right policy for resetting? Which model is right  here (stateful or stateless)?

![1_0](https://cloud.githubusercontent.com/assets/17245428/15579544/610f7c5c-2365-11e6-925f-e808c70c33fa.png)

![1_1](https://cloud.githubusercontent.com/assets/17245428/15579556/6571a86a-2365-11e6-8163-1ebd5af26bcc.png)
![1_2](https://cloud.githubusercontent.com/assets/17245428/15579558/68e4a81c-2365-11e6-94e8-3bae63f42354.png)
![1_3](https://cloud.githubusercontent.com/assets/17245428/15579560/6b529276-2365-11e6-83af-8a0e26e6c10a.png)

![1_4](https://cloud.githubusercontent.com/assets/17245428/15579677/fae3900c-2365-11e6-9604-ed2213d836df.png)

**Comment by lemuriandezapada** [comment]
in the stateful case your prediction will depend on internal state values (the initial states for each batch/last states from the previous batch) that will be changing over time. This might lead to some instability issues. 
When I tried it it worked better by setting the learning rate in stateful mode some order of magnitude lower. You also need to take into account that in stateful mode your model will always be a bit more biased towards the end of your sequence. 
You might also maybe want to try pretraining stateless and then moving to stateful for finetuning.

**Comment by mininaNik** [comment]
@lemuriandezapada Thank you
> setting the learning rate in stateful mode some order of magnitude lower
I don't understand what you mean. My optimizer is rmsprop and the default value for lr is 0.001

**Comment by gamehere007** [comment]
@mininaNik 

I have a question here. Why did you use two LSTM layers here?

**Comment by mininaNik** [comment]
@gamehere007  The first one is encoder and the second one is decoder.

**Comment by beloteloff** [comment]
Hi! Where is the 5 future values, which you wanted to predict? As I see on your plots your have only prediction "for past" and no future values on plots.
